### PR TITLE
`indices.query.bool.max_clause_count` now limits all query clauses

### PR DESCRIPTION
**Pull request by jpountz**
In the upcoming Lucene 9 release, `indices.query.bool.max_clause_count` is
going to apply to the entire query tree rather than per `bool` query. In order
to avoid breaks, the limit has been bumped from 1024 to 4096.

The semantics will effectively change when we upgrade to Lucene 9, this PR
is only about agreeing on a migration strategy and documenting this change.

To avoid further breaks, I am leaning towards keeping the current setting name
even though it contains `bool`. I believe that it still makes sense given that
`bool` queries are typically the main contributors to high numbers of clauses.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/SearchModule.java [hunk 1]
```diff
@@ -263,7 +263,7 @@
  */
 public class SearchModule {
     public static final Setting<Integer> INDICES_MAX_CLAUSE_COUNT_SETTING = Setting.intSetting("indices.query.bool.max_clause_count",
-            1024, 1, Integer.MAX_VALUE, Setting.Property.NodeScope);
+            2048, 1, Integer.MAX_VALUE, Setting.Property.NodeScope);
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/search/SearchModule.java:266]: I think this should be 4096?

File: docs/reference/modules/indices/search-settings.asciidoc [hunk 2]
```diff
@@ -7,16 +7,16 @@ limits.
...
-to performance degradations and memory issues, especially in clusters with a high load or 
-few resources.
+This setting limits the total number of clauses that a query tree can have. The default of 4096
+is quite high and should normally be sufficient. This limit applies to the rewritten query, so
+not only `bool` queries can contribute high numbers of clauses, but also all queries that rewrite
+to `bool` queries internally such as `fuzzy` queries. The limit is in place to prevent searches
+from becoming too large, and taking up too much CPU and memory. In case you're considering
+increasing this setting, make sure you've exhausted all other options to avoid having to do this.
+Higher values can lead to performance degradations and memory issues, especially in clusters with
+a high load or few resources.
```
Comment (romseygeek) [inline docs/reference/modules/indices/search-settings.asciidoc:19]: I wonder if it's worth mentioning `terms` queries and `index_prefix` settings here as well?

**Review by romseygeek — COMMENTED** [review]
Looks good! I have a couple of questions.

File: server/src/main/java/org/elasticsearch/search/SearchModule.java (hunk 1 above)
Comment (jpountz) [reply server/src/main/java/org/elasticsearch/search/SearchModule.java]: I went back and forth between 2048 and 4096, thanks for catching!

File: docs/reference/migration/migrate_8_0/search.asciidoc [hunk 3]
```diff
@@ -20,6 +20,22 @@ Aggregating and sorting on `_id` should be avoided. As an alternative, the
 `_id` field's contents can be duplicated into another field with docvalues
 enabled (note that this does not apply to auto-generated IDs).
 ====
+
+[[max_clause_count_change]]
+.Changed semantics of the `indices.query.bool.max_clause_count` setting.
```
Comment (jrodewig) [inline docs/reference/migration/migrate_8_0/search.asciidoc:25]:
```suggestion
.The `indices.query.bool.max_clause_count` setting now limits all query clauses.
```

**Review by jrodewig — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**